### PR TITLE
Core 1723 infura websockets

### PR DESCRIPTION
**Pull request by isentropy**
This PR allows GetEvents to poll event data over websockets instead of http. This is necessary to get event data from Infura nodes, which only support filters over websockets. Here are the major changes:

- I renamed EthereumJsonRpc as HttpEthereumJsonRpc and created a new class WebsocketsEthereumJsonRpc, Both HttpEthereumJsonRpc and WebsocketsEthereumJsonRpc  extend a common abstract base class EthereumJsonRpc. WebsocketsEthereumJsonRpc and HttpEthereumJsonRpc can be easily interchanged.
- rpcCall() no longer returns the RPC response, which is instead handled by an interface JsonRpcResponseHandler that is passed to the EthereumJsonRpc constructor. In the case of WebsocketsEthereumJsonRpc, the handler is called by another thread when the response arrives. In the case of HttpEthereumJsonRpc, the handler is invoked directly by rpcCall() in the same thread. This allows for both sync and async calls.
- ID's are actually used non-trivially now when sending RPC request, and the JsonRpcResponseHandler looks at the ID of the response to figure out which request it corresponds to.
- GetEvents first tries to get the websocket connection by looking at ("streamr.ethereum.wss." + network) config param. If it's not there, it falls back to http.

ContractEventPoller implements JsonRpcResponseHandler and GetEvents implements EventsListener (formerly ContractEventPoller.Listener), which follows only event logs and errors.

### CODE REVIEW

**Comment by hpihkala** [comment]
Didn't read through the changes properly yet, but an immediate question:

- What's the connection handling like? Has its robustness against disconnections etc. been considered? Or does the underlying websocket library provide basic robustness (like, will it reconnect if the connection is broken?) Will the `GetEvents` module recover from a disconnect-reconnect sequence?

**Comment by isentropy** [comment]
Yes it should handle disconnects, but I think I need to implement anError method. It's my first time programming websocket. I will test. This is the standard javax.websockets package.

**Comment by isentropy** [comment]
I tried disconnecting my internet while WebsocketsEthereumJsonRpc was running.  Interestingly, the websocket enpoint never received an onError message. Outgoing messages didn't complete and there were no incoming messages, but it **did recover when I resumed connection** a few minutes later. Looks like you can call get(long time) on the send request to see if it times out, but otherwise there was no indication of disconnect.

File: src/java/com/unifina/signalpath/blockchain/ContractEventPoller.java [hunk 1]
```diff
@@ -77,30 +86,41 @@ private void newFilter() {
...
 			filterId, contractAddress, id));
-	}
 
-	/**
-	 * https://github.com/ethereum/wiki/wiki/JSON-RPC#eth_getfilterchanges
-	 */
+	}
 	private synchronized void pollChanges() {
-		log.info(String.format("Polling filter '%s'.", filterId));
+		if(filterId == null){
```
Comment (kare) [inline src/java/com/unifina/signalpath/blockchain/ContractEventPoller.java]: Please pay attention to formatting. IDEA is your friend here:
- Select method
- Code -> Reformat code

File: src/java/com/unifina/signalpath/blockchain/ContractEventPoller.java [hunk 2]
```diff
@@ -77,30 +86,41 @@ private void newFilter() {
...
 		try {
-			JSONObject response = rpc.rpcCall("eth_getFilterChanges", singletonList(filterId));
-			JSONArray jsonArray = response.getJSONArray("result");
-			if (jsonArray.length() != 0) {
-				listener.onEvent(jsonArray);
-			}
-		} catch (EthereumJsonRpc.ErrorObjectError e) {
+			log.info(String.format("Polling filter '%s'.", filterId));
+			rpc.rpcCall("eth_getFilterChanges", singletonList(filterId), ID_POLLFILTER);
+		}
```
Comment (kare) [inline src/java/com/unifina/signalpath/blockchain/ContractEventPoller.java]: Formatting

File: src/java/com/unifina/signalpath/blockchain/ContractEventPoller.java [hunk 3]
```diff
@@ -77,30 +86,41 @@ private void newFilter() {
...
 				filterId = null;
 				newFilter();
 			} else {
 				listener.onError(e.getMessage());
 			}
-		} catch (EthereumJsonRpc.Error | JSONException e) {
+		} catch (HttpEthereumJsonRpc.RPCException | JSONException e) {
 			listener.onError(e.getMessage());
 		}
+		catch(Exception e){
```
Comment (kare) [inline src/java/com/unifina/signalpath/blockchain/ContractEventPoller.java]: Formatting

File: src/java/com/unifina/signalpath/blockchain/ContractEventPoller.java [hunk 4]
```diff
@@ -114,20 +134,42 @@ private void uninstallFilter() {
 		filterId = null;
 
 		try {
-			result = rpc.rpcCall("eth_uninstallFilter", singletonList(id)).getBoolean("result");
-		} catch (JSONException e) {
+			rpc.rpcCall("eth_uninstallFilter", singletonList(id), ID_REMOVEFILTER);
+		}
```
Comment (kare) [inline src/java/com/unifina/signalpath/blockchain/ContractEventPoller.java]: Formatting

File: src/java/com/unifina/signalpath/blockchain/ContractEventPoller.java [hunk 5]
```diff
@@ -114,20 +134,42 @@ private void uninstallFilter() {
...
 
 	private static boolean filterDoesNotExist(int code) {
 		return code == -32000;
 	}
+
+	@Override
+	public void processResponse(JSONObject resp) {
+		int id = resp.getInt("id");
+		switch(id){
+			case ID_ADDFILTER: processAddFilterResponse(resp); return;
```
Comment (kare) [inline src/java/com/unifina/signalpath/blockchain/ContractEventPoller.java]: Should default: case be here?

File: src/java/com/unifina/signalpath/blockchain/EthereumJsonRpc.java [hunk 6]
```diff
@@ -2,85 +2,33 @@
...
-			"id", callId,
-			"jsonrpc", "2.0",
-			"method", method,
-			"params", params
-		)));
-	}
+	/*
+		implementation should call handler.processResponse()
+	 */
+	public abstract void rpcCall(String method, List params, int callId) throws Exception;
```
Comment (kare) [inline src/java/com/unifina/signalpath/blockchain/EthereumJsonRpc.java]: More specific exception type?

File: src/java/com/unifina/signalpath/blockchain/GetEvents.java [hunk 7]
```diff
@@ -84,9 +86,17 @@ public void initialize() {
 
 	@Override
 	public void onStart() {
-		String rpcUrl = ethereumOptions.getRpcUrl();
+		String rpcUrl = ethereumOptions.getWebsocketRpcUri();
+		if(rpcUrl == null){
```
Comment (kare) [inline src/java/com/unifina/signalpath/blockchain/GetEvents.java]: Formatting

File: src/java/com/unifina/signalpath/blockchain/HttpEthereumJsonRpc.java [hunk 8]
```diff
@@ -0,0 +1,70 @@
...
+	}
+
+	abstract class RPCException extends Exception {
+		private RPCException(String message) {
+			super(String.format("JSON RPC error with server '%s': %s", url, message));
+		}
+	}
+
+
+	class HttpStatusException extends RPCException {
```
Comment (kare) [inline src/java/com/unifina/signalpath/blockchain/HttpEthereumJsonRpc.java]: static

File: src/java/com/unifina/signalpath/blockchain/HttpEthereumJsonRpc.java [hunk 9]
```diff
@@ -0,0 +1,70 @@
...
+		private HttpStatusException(int statusCode, JsonNode body) {
+			super(String.format("unexpected status code %d with content %s", statusCode, body));
+		}
+	}
+
+	class ErrorObjectException extends RPCException {
```
Comment (kare) [inline src/java/com/unifina/signalpath/blockchain/HttpEthereumJsonRpc.java]: static

File: src/java/com/unifina/signalpath/blockchain/JsonRpcResponseHandler.java [hunk 10]
```diff
@@ -0,0 +1,11 @@
+package com.unifina.signalpath.blockchain;
+
+import org.json.JSONObject;
+
+/*
+processes raw JSON response
+ */
+
+public interface JsonRpcResponseHandler {
+	public void processResponse(JSONObject resp);
```
Comment (kare) [inline src/java/com/unifina/signalpath/blockchain/JsonRpcResponseHandler.java]: public is unnecessary.

File: src/java/com/unifina/signalpath/blockchain/WebsocketEthereumJsonRpc.java [hunk 11]
```diff
@@ -0,0 +1,71 @@
...
+import org.json.JSONArray;
+import org.json.JSONObject;
+
+public class WebsocketEthereumJsonRpc extends EthereumJsonRpc {
+	private static final Logger log = Logger.getLogger(WebsocketEthereumJsonRpc.class);
+	private EthereumRpcEndpoint wsEndpoint;
+	Session userSession = null;
+
+	@ClientEndpoint
+	protected class EthereumRpcEndpoint{
```
Comment (kare) [inline src/java/com/unifina/signalpath/blockchain/WebsocketEthereumJsonRpc.java]: static

File: src/java/com/unifina/signalpath/blockchain/WebsocketEthereumJsonRpc.java [hunk 12]
```diff
@@ -0,0 +1,71 @@
+package com.unifina.signalpath.blockchain;
+
+import com.google.gson.JsonParser;
+import org.apache.log4j.Logger;
+import javax.websocket.*;
+import javax.websocket.Session;
+import java.io.IOException;
```
Comment (kare) [inline src/java/com/unifina/signalpath/blockchain/WebsocketEthereumJsonRpc.java]: Sort imports. You can do it automatically with IDEA:
- Preferences -> Editor -> Auto Import -> [x] Optimize imports on the fly

File: src/java/com/unifina/signalpath/blockchain/WebsocketEthereumJsonRpc.java [hunk 13]
```diff
@@ -0,0 +1,71 @@
...
+import java.util.List;
+
+import com.google.gson.JsonObject;
+import org.json.JSONArray;
+import org.json.JSONObject;
+
+public class WebsocketEthereumJsonRpc extends EthereumJsonRpc {
+	private static final Logger log = Logger.getLogger(WebsocketEthereumJsonRpc.class);
+	private EthereumRpcEndpoint wsEndpoint;
+	Session userSession = null;
```
Comment (kare) [inline src/java/com/unifina/signalpath/blockchain/WebsocketEthereumJsonRpc.java]: userSession could be private?

File: src/java/com/unifina/signalpath/blockchain/HttpEthereumJsonRpc.java [hunk 14]
```diff
@@ -0,0 +1,70 @@
...
+
+import java.util.List;
+
+class HttpEthereumJsonRpc extends EthereumJsonRpc{
+	HttpEthereumJsonRpc(String url, JsonRpcResponseHandler handler) {
+		super(url, handler);
+	}
+
+	@Override
+	public void rpcCall(String method, List params, int callId) throws HttpStatusException, ErrorObjectException, com.mashape.unirest.http.exceptions.UnirestException {
```
Comment (kare) [inline src/java/com/unifina/signalpath/blockchain/HttpEthereumJsonRpc.java]: How about "throws RPCCallException"?

File: src/java/com/unifina/signalpath/blockchain/WebsocketEthereumJsonRpc.java [hunk 15]
```diff
@@ -0,0 +1,71 @@
...
+		open();
+	}
+
+	@Override
+	public void rpcCall(String method, List params, int callId){
+		wsEndpoint.sendMessage(formRequestBody(method, params, callId));
+	}
+
+
+	public void open() throws URISyntaxException, IOException, DeploymentException {
```
Comment (kare) [inline src/java/com/unifina/signalpath/blockchain/WebsocketEthereumJsonRpc.java]: How about "throws RPCCallException"?

**Comment by isentropy** [comment]
@hpihkala Reconnection is working. I forced disconnection by turning internet off and after about 20m the endpoint receives an onClose (not onError) message. You can see if it's a disconnect or normal closure by checking the CloseReason code.  In the case of an abnormal closure (or orError) it attempts a reconnect.

File: src/java/com/unifina/signalpath/blockchain/EthereumJsonRpc.java (hunk 6 above)
Comment (isentropy) [reply src/java/com/unifina/signalpath/blockchain/EthereumJsonRpc.java]: well the rpc could, in theory, go over a number of channels: http, websocket, unix socket file, etc. Specific implementations throw different exceptions (eg HttpException). So I would keep the Exception generic.

File: src/java/com/unifina/signalpath/blockchain/WebsocketEthereumJsonRpc.java (hunk 15 above)
Comment (isentropy) [reply src/java/com/unifina/signalpath/blockchain/WebsocketEthereumJsonRpc.java]: eh, this looks fine to me. Malformed URI should be a different exception from connection issue IMO

**Comment by isentropy** [comment]
@hpihkala can I merge this?

**Pull request merged**